File: src/main/java/com/team3/gudit/outbox/publisher/OutboxPublisher.java
package com.team3.gudit.outbox.publisher;

import com.team3.gudit.outbox.entity.OutboxEvent;
import com.team3.gudit.outbox.entity.OutboxEventStatus;
import com.team3.gudit.outbox.repository.OutboxEventRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.redis.connection.stream.MapRecord;
import org.springframework.data.redis.connection.stream.StreamRecords;
import org.springframework.data.redis.core.StringRedisTemplate;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Map;

@Slf4j
@Component
@RequiredArgsConstructor
public class OutboxPublisher {

    private final OutboxEventRepository outboxEventRepository;
    private final StringRedisTemplate stringRedisTemplate;

    @Scheduled(fixedDelay = 1000)
    @Transactional
    public void publishPendingEvents() {
        List<OutboxEvent> events =
                outboxEventRepository
                        .findAllByStatusOrderByCreatedAtAsc(
                                OutboxEventStatus.PENDING
                        );

        for (OutboxEvent event : events) {
            publish(event);
        }
    }

    private void publish(OutboxEvent event) {
        try {
            MapRecord<String, String, String> record =
                    StreamRecords
                            .newRecord()
                            .in(
                                    StockRestoreStreamConstants
                                            .STOCK_RESTORE_STREAM
                            )
                            .ofMap(
                                    Map.of(
                                            "eventId",
                                            event.getEventId(),

                                            "traceId",
                                            event.getTraceId() == null
                                                    ? ""
                                                    : event.getTraceId(),

                                            "eventType",
                                            event.getEventType().name(),

                                            "payload",
                                            event.getPayload()
                                    )
                            );

            stringRedisTemplate
                    .opsForStream()
                    .add(record);

            event.markPublished();

            log.info(
                    "Outbox event published. "
                            + "eventId={}, eventType={}",
                    event.getEventId(),
                    event.getEventType()
            );

        } catch (RuntimeException e) {
            log.warn(
                    "Outbox event publish failed. "
                            + "eventId={}, eventType={}",
                    event.getEventId(),
                    event.getEventType(),
                    e
            );
        }
    }
}

File: src/main/kotlin/com/team3/gudit/outbox/publisher/OutboxPublisher.kt
package com.team3.gudit.outbox.publisher

import com.team3.gudit.outbox.entity.OutboxEvent
import com.team3.gudit.outbox.entity.OutboxEventStatus
import com.team3.gudit.outbox.entity.OutboxEventType
import com.team3.gudit.outbox.repository.OutboxEventRepository
import org.slf4j.LoggerFactory
import org.springframework.data.redis.connection.stream.MapRecord
import org.springframework.data.redis.connection.stream.StreamRecords
import org.springframework.data.redis.core.StringRedisTemplate
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional

@Component
class OutboxPublisher(
    private val outboxEventRepository: OutboxEventRepository,
    private val stringRedisTemplate: StringRedisTemplate
) {

    private val log = LoggerFactory.getLogger(javaClass)

    @Scheduled(fixedDelay = 1000)
    @Transactional
    fun publishPendingEvents() {
        val events =
            outboxEventRepository.findAllByStatusOrderByCreatedAtAsc(
                OutboxEventStatus.PENDING
            )

        for (event in events) {
            publish(event)
        }
    }

    private fun publish(event: OutboxEvent) {
        try {
            val record: MapRecord<String, String, String> =
                StreamRecords
                    .newRecord()
                    .`in`(resolveStream(event))
                    .ofMap(
                        mapOf(
                            "eventId" to event.eventId,
                            "traceId" to (event.traceId ?: ""),
                            "eventType" to event.eventType.name,
                            "payload" to event.payload
                        )
                    )

            stringRedisTemplate
                .opsForStream<String, String>()
                .add(record)

            event.markPublished()

            log.info(
                "Outbox event published. eventId={}, eventType={}",
                event.eventId,
                event.eventType
            )
        } catch (e: RuntimeException) {
            log.warn(
                "Outbox event publish failed. eventId={}, eventType={}",
                event.eventId,
                event.eventType,
                e
            )
        }
    }

    private fun resolveStream(event: OutboxEvent): String =
        when (event.eventType) {
            OutboxEventType.STOCK_RESTORE_REQUESTED ->
                StockRestoreStreamConstants.STOCK_RESTORE_STREAM

            OutboxEventType.PAYMENT_COMPENSATION_REQUIRED ->
                PaymentCompensationStreamConstants.PAYMENT_COMPENSATION_STREAM
        }
}